src/metrics/carbon/parser.go

```diff
@@ -0,0 +1,289 @@
+// Copyright (c) 2019 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package carbon
+
+import (
+	"bufio"
+	"errors"
+	"fmt"
+	"io"
+	"math"
+	"strconv"
+	"strings"
+	"time"
+	"unicode/utf8"
+
+	"github.com/m3db/m3x/unsafe"
+)
+
+const (
+	negativeNanStr = "-nan"
+	nanStr         = "nan"
+
+	floatFormatByte = 'f'
+	floatPrecision  = -1
+	floatBitSize    = 64
+	intBase         = 10
+)
+
+var (
+	errInvalidLine      = errors.New("invalid line")
+	errNotUTF8          = errors.New("not valid UTF8 string")
+	recvValidTimeBuffer = time.Hour
+	mathNan             = math.NaN()
+)
+
+// Metric represents a carbon metric.
+type Metric struct {
+	Name []byte
+	Time time.Time
+	Val  float64
+	Recv time.Time
+}
+
+// ToLine converts the carbon Metric struct to a line.
+func (m *Metric) ToLine() string {
+	return string(m.Name) + " " + strconv.FormatFloat(m.Val, floatFormatByte, floatPrecision, floatBitSize) +
+		" " + strconv.FormatInt(m.Time.Unix(), intBase) + "\n"
+}
+
+// ParsePacket parses a carbon packet and returns the metrics and number of malformed lines.
+func ParsePacket(packet []byte) ([]Metric, int) {
+	var malformed, prevIdx, i int
+	mets := []Metric{}
```

Comment:
Maybe make this a param to pass in?  So we can pool the slices.  

Can split up into two methods to make both the nice and the efficient APIs available:
```
func ParsePacket(packet []byte) ([]Metric, int)
func AppendAndParsePacket(slice []Metric, packet []byte) ([]Metric, int)
```